### PR TITLE
Cache issues and issues count queries

### PR DESCRIPTION
**Pull request by Biboswan**
Apologies for the massive delay. Got busy with other stuff
Have a look @alexdor 
I wonder will it really help if we precache the `/issues?*`  results when `/issues-count` endpoint is hit in a go routine. 

I will add the docker configuration as well.

### CODE REVIEW

File: database.yml [hunk 1]
```diff
@@ -3,7 +3,7 @@ development:
   database: fixme_backend_development
   user: postgres
   password: makis123
-  url: {{ envOr "DATABASE_URL" "postgres://postgres:postgres@127.0.0.1:5432/fixme_backend_development?sslmode=disable"}}
```
Comment (alexdor) [inline database.yml]: Could you revert this change because it's going to break the production env?

File: worker/worker.go [hunk 2]
```diff
@@ -401,6 +402,11 @@ func (w *Worker) parseAndSaveIssues(issueData issueQueryWithBefore, repository *
 		fmt.Println(errors.WithMessage(err, "failed to update issues"))
 	}
 
+	cacheConn := cache.CachePool.Get()
+	defer cacheConn.Close()
+
+	cache.DeleteAllKeys(&cacheConn)
```
Comment (alexdor) [inline worker/worker.go]: Deleting everything might break things down the road.

 Could you try prefixing the cache keys (for example for pages you prefix the keys with `page_` and for count with `count_`) and instead of deleting everything only delete the keys that start `page_` and the keys that start with `count_`

File: actions/issues.go [hunk 3]
```diff
@@ -37,10 +46,38 @@ func (v IssuesResource) ListOpen(c buffalo.Context) error {
...
 	}
-	// Retrieve all Issues from the DB
-	if err := q.Where(whereClause).All(issues); err != nil {
-		return errors.WithStack(err)
+
+	page := params.Get("page")
+	redisKey := "issues:" + whereClause + "and page=" + page
+	ok, err := cache.Exists(&cacheConn, redisKey)
+	if err != nil {
+		fmt.Println(errors.WithMessage(err, "Cache operation failed"))
```
Comment (alexdor) [inline actions/issues.go]: If smth goes wrong here, could you fetch the issues from the db instead of only printing an error?

**Review by alexdor — CHANGES_REQUESTED** [review]
Thank you for the pr : ) We could precache both of them but even if we only precache the issues query it's better than nothing because a count query is much faster than retrieving all the data. 

I've noted a few changes, let me know what do you think about them : )

File: database.yml (hunk 1 above)
Comment (Biboswan) [reply database.yml]: my bad i missed that case

File: worker/worker.go (hunk 2 above)
Comment (Biboswan) [reply worker/worker.go]: currently, the keys are prefixed with issues: and issues-count

Comment (alexdor) [reply worker/worker.go]: You are right, my bad. Could you delete only the keys that have these prefixes instead of all the keys?

Comment (Biboswan) [reply worker/worker.go]: Sure! with some scan... since there is no direct command for such an operation i guess.

Comment (alexdor) [reply worker/worker.go]: Yeah, the suggested way (if I remember correctly) is to scan for the keys and then use the unlink command to drop all the keys that were returned from the scan

**Comment by Biboswan** [comment]
How shall we set some redis configuration like `maxmemory **MB`,`maxmemory-policy allkeys-lfu` 
 While initializing the cache in the app with redis client itself (don't think a nice idea) or set those parameters in the settings in the production. For development mode are these settings required?

**Comment by alexdor** [comment]
It's probably better to set them through the config file of redis rather than the client. I don't think that these settings are required for dev, only for production.

**Comment by Biboswan** [comment]
@alexdor i'm just confused over which way to go. If we use a `lua` script and scan there then it will be blocking. On the other hand if we implement the `scan` and `unlink` right in the go server. Then we shall use `ttl(key)` to find whether the key was set before the start of deletion process or is a very new key which is valid and shouldn't be deleted.

**Comment by alexdor** [comment]
I think that it’s ok to delete a key even if it’s a relative new key. And that it’s better to implement it in go so we don’t also have manage a lua script. Otherwise we will introduce unnecessary complexity, what do you think?

**Comment by Biboswan** [comment]
Yup keeping things in go.

**Comment by Biboswan** [comment]
Incorporated the suggestions and and also added docker

**Comment by Biboswan** [comment]
While running the docker `Server` should be `cache` here https://github.com/Biboswan/fixme_backend/blob/cache_layer/cache/cache.go#L14. If this is the best way to go then shall tell so in the Readme

File: actions/issues.go [hunk 4]
```diff
@@ -111,10 +162,37 @@ func (v IssuesResource) Count(c buffalo.Context) error {
...
 	}
-	count, err := q.Where(whereClause).Count(issues)
-	// Count Issues from the DB
+
+	page := params.Get("page")
+	cacheKey := "issues-count:" + whereClause + "and page=" + page
+	ok, err := cache.Exists(&cacheConn, cacheKey)
 	if err != nil {
-		return errors.WithStack(err)
+		fmt.Println(errors.WithMessage(err, "Cache operation failed"))
```
Comment (alexdor) [inline actions/issues.go]: we should also get the issues from the db if smth goes wrong here.

You can try refactoring this part to smth like this:

```

page := params.Get("page")
cacheKey := "issues-count:" + whereClause + "and page=" + page
ok, err := cache.Exists(&cacheConn, cacheKey)
if err != nil || !ok {
    count, err = q.Where(whereClause).Count(issues)
    // Count Issues from the DB
    if err != nil {
      return errors.WithStack(err)
    }

    _, err = cache.SetEx(&cacheConn, cacheKey, 600, count)
    if err != nil {
      fmt.Println(errors.WithMessage(err, "Cache operation failed"))
    }
} else {
    count, err = cache.GetInt(&cacheConn, cacheKey)

    if err != nil {
      fmt.Println(errors.WithMessage(err, "Cache operation failed"))
      count, err = q.Where(whereClause).Count(issues)
      // Count Issues from the DB
      if err != nil {
        return errors.WithStack(err)
    }
  } 
}
	
```

File: actions/issues.go [hunk 5]
```diff
@@ -37,10 +48,45 @@ func (v IssuesResource) ListOpen(c buffalo.Context) error {
...
-	// Retrieve all Issues from the DB
-	if err := q.Where(whereClause).All(issues); err != nil {
-		return errors.WithStack(err)
+
+	page := params.Get("page")
+	cacheKey := "issues:" + whereClause + "and page=" + page
+
+	ok, err := cache.Exists(&cacheConn, cacheKey)
+	if err != nil {
+		fmt.Println(errors.WithMessage(err, "Cache operation failed"))
```
Comment (alexdor) [inline actions/issues.go]: Could you also refactor this part a bit like the suggested refactor on the count cache?

File: actions/issues.go [hunk 6]
```diff
@@ -37,10 +48,45 @@ func (v IssuesResource) ListOpen(c buffalo.Context) error {
...
+			_, err := cache.SetEx(&cacheConn, cacheKey, 600, string(jsonIssues))
+			if err != nil {
+				fmt.Println(errors.WithMessage(err, "Cache operation failed"))
+			}
+
+		}
 	}
+
+	//Caching issues of next page of the same query
+	go preCacheIssues(whereClause, params, page)
```
Comment (alexdor) [inline actions/issues.go:80]: Nice touch : ) it's going to speed up things a lot

File: worker/worker.go [hunk 7]
```diff
@@ -401,6 +402,12 @@ func (w *Worker) parseAndSaveIssues(issueData issueQueryWithBefore, repository *
 		fmt.Println(errors.WithMessage(err, "failed to update issues"))
 	}
 
+	cacheConn := cache.CachePool.Get()
+	defer cacheConn.Close()
+
+	cache.DeleteKeysByPattern(&cacheConn, "issues:*")
+	cache.DeleteKeysByPattern(&cacheConn, "issues-count:*")
+
```
Comment (alexdor) [inline worker/worker.go:408]: Could you also precache the home page after the deletion of this 2?

File: actions/issues.go [hunk 8]
```diff
@@ -149,3 +227,42 @@ func requestParamToQueryFilter(query, paramValue, paramName *string) {
...
+	nextPageStr := strconv.Itoa(nextPage)
+	nextParams := params.(url.Values)
+	nextParams.Set("page", nextPageStr)
+	nextQ := tx.PaginateFromParams(nextParams).Eager()
+
+	nextCacheKey := "issues:" + whereClause + "and page=" + nextPageStr
+
+	ok, _ := cache.Exists(&cacheConn, nextCacheKey)
+
+	if ok == false {
```
Comment (alexdor) [inline actions/issues.go]:
```suggestion
	if !ok {
```

File: actions/issues.go [hunk 9]
```diff
@@ -149,3 +227,42 @@ func requestParamToQueryFilter(query, paramValue, paramName *string) {
...
+		if err := nextQ.Where(whereClause).All(issues); err != nil {
+			fmt.Println(errors.WithMessage(err, "DB Operation falied"))
+			fmt.Println(nextQ)
+			return
+		}
+
+		jsonIssues, _ := json.Marshal(issues)
+		_, err := cache.SetEx(&cacheConn, nextCacheKey, 600, string(jsonIssues))
```
Comment (alexdor) [inline actions/issues.go:248]: could you also add the count of the issues to cache?

File: cache/cache.go [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+	"log"
+	"time"
+
+	"github.com/caarlos0/env"
+	"github.com/gomodule/redigo/redis"
+	"github.com/pkg/errors"
+)
+
+type config struct {
+	Server      string        `env:"CACHE_SERVER" envDefault:"127.0.0.1:6379"`
```
Comment (alexdor) [inline cache/cache.go]: Could you rename the env variables here from `CACHE_` to `REDIS_` ? It's easier for a reader to understand which cache server is used

File: cache/cache.go [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+
+//SetEx sets a new key in to database and also expiry time in ms
+func SetEx(RConn *redis.Conn, key string, ttl int, data interface{}) (interface{}, error) {
+	return (*RConn).Do("SETEX", key, ttl, data)
+}
+
+//GetString returns the value for the key in string format
+func GetString(RConn *redis.Conn, key string) (string, error) {
+	//get
+	return redis.String((*RConn).Do("GET", key))
```
Comment (alexdor) [inline cache/cache.go:56]: Doesn't this also do the same?
```suggestion
	return redis.String(RConn.Do("GET", key))
```

**Review by alexdor — CHANGES_REQUESTED** [review]
Really nice work : ) I've noted a few small changes

**Comment by alexdor** [comment]
> While running the docker `Server` should be `cache` here https://github.com/Biboswan/fixme_backend/blob/cache_layer/cache/cache.go#L14. If this is the best way to go then shall tell so in the Readme

I'm not sure that I understood what you mean here

**Comment by Biboswan** [comment]
> > While running the docker `Server` should be `cache` here https://github.com/Biboswan/fixme_backend/blob/cache_layer/cache/cache.go#L14. If this is the best way to go then shall tell so in the Readme
> 
> I'm not sure that I understood what you mean here

While running the docker `CACHE_SERVER`variable in the stated line should be set to value`cache` (see docker-compose.yml )for the docker compose to resolve the cache (hostname) to IP address of the cache container.

**Comment by alexdor** [comment]
Yeah, I'll create a link between Redis and the api and set the env variables inside compose to reflect the location of Redis before I deploy it, no worries : )

File: actions/issues.go (hunk 4 above)
Comment (Biboswan) [reply actions/issues.go]: Acc to the above snippet we miss to print the cache error if `cacheConn.Exists` throws error.

File: cache/cache.go (hunk 11 above)
Comment (Biboswan) [reply cache/cache.go:56]: I think they are different since  get this err `RConn.Do undefined (type *redis.Conn is pointer to interface, not interface)`

File: actions/issues.go (hunk 4 above)
Comment (Biboswan) [reply actions/issues.go]: but yes the if else is scary

File: actions/issues.go (hunk 8 above)
Comment (Biboswan) [reply actions/issues.go]: actually I did this what if ok  = `nil`. is !nil allowed?

Comment (Biboswan) [reply actions/issues.go]: oops ok is never nil

File: actions/issues.go (hunk 4 above)
Comment (alexdor) [reply actions/issues.go]: yeah the idea is to actually create a logger package so we will send the error to the logger and he will ignore it when it's nill. So a simple `TODO` comment should be sufficient to add the error printing back later. The main issue with the nested if else is that we are rewriting the same code twice which could lead into strange corner cases

File: actions/issues.go (hunk 9 above)
Comment (Biboswan) [reply actions/issues.go:248]: Do you mean the total count of issues for the query not paged result (issues-count) ?

Comment (alexdor) [reply actions/issues.go:248]: Sorry my suggestion doesn't make sense, I'll remove it. My bad

File: worker/worker.go [hunk 12]
```diff
@@ -493,4 +498,45 @@ func (w *Worker) searchForDanglingIssues(repository *models.Repository) {
...
+	q := tx.PaginateFromParams(params).Eager()
+	ok, _ := cache.Exists(&cacheConn, cacheKey)
+
+	if !ok {
+		if err := q.Where(defaultIssuesWhereClause).All(issues); err != nil {
+			fmt.Println(errors.WithMessage(err, "DB Operation falied"))
+			return
+		}
+
+		jsonIssues, _ := json.Marshal(issues)
```
Comment (alexdor) [inline worker/worker.go]: could you print the error here and return? If the issues could be marshalled they are not going to be parsed properly when a client tries to get them

**Review by alexdor — CHANGES_REQUESTED** [review]

**Comment by Biboswan** [comment]
@alexdor changes were made!

**Comment by alexdor** [comment]
@Biboswan I've managed to merge your branch into master 🎉 

There where a few issues, nothing major though. You shouldn't silently ignore the errors, you should always handle them (there was a crash for example when no page was provided because you didn't check if the string to int conversation had an error)

Once again, good work and thank you for your contribution : )

**Comment by Biboswan** [comment]
Thanks @alexdor  :) Happy to contribute
